src/java/org/apache/cassandra/db/partitions/UnfilteredPartitionIterators.java

```diff
@@ -132,17 +140,20 @@ protected UnfilteredRowIterator getReduced()
                                                                  ? null
                                                                  : listener.getRowMergeListener(partitionKey, toMerge);
 
-                // Make a single empty iterator object to merge, we don't need toMerge.size() copiess
-                UnfilteredRowIterator empty = null;
-
-                // Replace nulls by empty iterators
-                for (int i = 0; i < toMerge.size(); i++)
+                if (listener != null && listener.preserveOrder())
```

Comment:
We could also skip the empty iterators if `rowListener` turns out to be null.

As this is a bit more involved (getting this listener and setting up the array needs to happen on the first `reduce` call), feel free to ignore this comment.

Reply:
yeah going ignore this one for now. I was trying todo that initially when creating the pull request but due to complexity I skipped it.